### ISSUE TITLE
Proposal: reconsider CLSCompliant and CA1014

### ISSUE THREAD

**Issue by huoyaoyuan** [dotnet/runtime#44194]
## Background

As far as I know, CLS was from the early versions of CLI standard, to mark members that "may not be supported by compilers that only supports the minimal set". Nowadays, it's causing confusion among the .NET ecosystem.

### False negative (marked with `[CLSCompliant(false)]`, but widely supported)

The most common non-CLS-complaint feature should be unsigned integers. Nowadays, all the three major alive languages (C#, VB, F#) support unsigned integers. Continuously marking them with `[CLSCompliant(false)]` is meaningless work at the sight of major consumers.

### False positive (marked with `[CLSCompliant(true)]`, but requires certain compiler to support)

As the runtime evolving, it introduces new features that's not defined in early versions of standards. The most famous one should be `Span<T>` and ref structs. **Howover, `Span<T>` is not marked with `[CLSCompliant(false)]`, but with `modreq`.** This makes `CLSCompiaint` a rigid mark that doesn't even fit older standard.

### Analyzer about "mark assembly with [CLSCompliant]" (CA1014) confuses developers

Although the BCL may still mark with `[CLSCompliant]` for standard conformant, library and application developers typically won't want to be bothered by it, because it means nothing in realistic.

As .NET 5 bringing NetAnalyzers inbox, it's bothering in more ways. Some developers (for example me) choose to enable `AnalysisMode = AllEnabledByDefault` for more strict analysis, and that brings CA1014.

There are certain legacy analyzers shipped with NetAnalyzers, but not included in `AllEnabledByDefault` (CA1021, CA1045). I think CA1014 may be included in it.

## Proposed actions

- Remove CA1014 from `AnalysisLevel_5_AllEnabledByDefault.editorconfig`, as it's not meaningful for .NET 5 applications, and not easy to turn off(dotnet/roslyn-analyzers#4400)
- Reconsider what to do with old CLS compliance in BCL
- If you decide to keep [CLSComplaint], consider the chance of marking `Span` and related apis with `[CLSCompliant(false)]`.

**Comment by carlossanlop** [comment]
> Remove CA1014 from AnalysisLevel_5_AllEnabledByDefault.editorconfig, as it's not meaningful for .NET 5 applications, and not easy to turn off(dotnet/roslyn-analyzers#4400)
> Reconsider what to do with old CLS compliance in BCL
> If you decide to keep [CLSComplaint], consider the chance of marking Span and related apis with [CLSCompliant(false)].

@terrajobst @stephentoub @mavasani do you mind sharing your opinion on these proposed actions?

**Comment by mavasani** [comment]
Yes, this makes complete sense and we have had multiple reports for it. I was planning to do this soon. Feel free to port it to analyzers repo and assign it to me.

**Comment by carlossanlop** [comment]
> Yes, this makes complete sense and we have had multiple reports for it.

Ok, thanks for the confirmation.

> I was planning to do this soon. Feel free to port it to analyzers repo and assign it to me.

This is the issue for the proposal, and it needs to go through API review first. I'll ping you / assign this to you once it gets approved.

**Comment by bartonjs** [comment]
Removing the rule from AllEnabledByDefault, if possible, is reasonable.  But otherwise this is tightly coupled to the still in progress question of what we're doing with CLS compliance and we can't really take action at this time.

**Comment by RehanSaeed** [comment]
Corefx removed `<CLSCompliant>false</CLSCompliant>` from the BCL in https://github.com/dotnet/corefx/pull/9052. I'd love to find out more about why they did that.

CA1014 can now be turned off in .editorconfig in https://github.com/dotnet/roslyn-analyzers/issues/4400. I'd like to understand if thats the recommended approach today.

**Comment by mavasani** [comment]
> CA1014 can now be turned off in .editorconfig in https://github.com/dotnet/roslyn-analyzers/issues/4400. I'd like to understand if thats the recommended approach today.

@RehanSaeed This should be a simple change. Let me create a PR right away.

**Comment by stephentoub** [comment]
> Corefx removed <CLSCompliant>false</CLSCompliant> from the BCL in https://github.com/dotnet/corefx/pull/9052. I'd love to find out more about why they did that.

Because it wasn't actually necessary in those cases.

**Comment by tannergooding** [comment]
I don't believe we have any plans to change or otherwise "improve" `CLSCompliant`. It exists and will likely stay as it is without some further driving function from the language or runtime.
